low-level/framework/utils/sysfs_interface.py

```diff
@@ -138,7 +140,144 @@ def fetch_nw_cable_status(self, nw_interface_path, interface):
                 carrier_indicator = cFile.read().strip()
             if carrier_indicator not in self.nw_phy_link_state.keys():
                 carrier_indicator = 'unknown'
-        except OSError as os_err:
-            return os_err.errno
+        except Exception:
+            carrier_indicator = 'unknown'
         return self.nw_phy_link_state[carrier_indicator]
 
+    def fetch_nw_operstate(self, interface):
+        """
+        Return the Operational Status of the Network Interface.
+
+        Possible Values: <'UNKNOWN', 'NOTPRESENT", 'DOWN', 'LOWERLAYERDOWN',
+                          'TESTING', 'DORMANT', 'UP'>
+        """
+        operstate_file_path = os.path.join(self.get_sys_dir_path('net'),
+                                           interface, 'operstate')
+        operstate = "UNKNOWN"
+        try:
+            if os.path.isfile(operstate_file_path):
+                with open(operstate_file_path) as cFile:
+                    operstate = cFile.read().strip().upper()
+        except Exception as err:
+            raise err
+        return operstate
+
+    def get_sas_host_list(self):
+        """
+            Return SAS host list from /sys/class/sas_host directory.
+
+            eg: ['host1', 'host2']
+        """
+        sas_hosts = []
+        try:
+            host_dirpath = self.get_sys_dir_path('sas_host')
+            sas_hosts = os.listdir(host_dirpath)
+            sas_hosts = [host for host in sas_hosts
+                         if 'host' in host]
+        except Exception:
+            return []
+        return sas_hosts
+
+    def get_sas_port_list(self, host=""):
+        """
+            Return a list of SAS ports.
+
+            searches for ports from given host in /sys/class/sas_port
+            directory if directory exists else return an empty list.
+            eg: ['port-1:0', 'port-1:1', 'port-1:2', 'port-1:3']
+        """
+        sas_ports = []
+        try:
+            host_id = host.replace('host', '')
+            port_dirpath = self.get_sys_dir_path('sas_port')
+            sas_ports = os.listdir(port_dirpath)
+            sas_ports = [port for port in sas_ports
+                         if f'port-{host_id}' in port]
+        except Exception:
+            return []
```

Comment:
same here, reset sas_ports = [], so one return statement @ line 198 would suffice

Reply:
updated in new PR.